0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,406 @@
+
+When a connector receives a request to send a message to the peer, the raw message from the settlement engine MUST be encoded within an ILP PREPARE packet as described below. Then, the ILP PREPARE should be sent to the peer's connector associated with the account using ILP-over-HTTP.
+
+The peer's connector MUST extract the message data and forward it to its settlement engine for processing, and MUST NOT forward the ILP packet to any other connectors. When the settlement engine responds with a response message, the connector MUST encode the raw message from the settlement engine within an ILP FULFILL or ILP REJECT, depending upon the code of the response. If the connector was unable to process the request, it MUST respond with an ILP REJECT.
+
+##### PREPARE
+
+- `amount`: `0`
+- `expiresAt`: _Determined by connector_
+- `executionCondition`: `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`
+- `destination`: `peer.settle`
+- `data`: _Request message from sender settlement engine_
+
+##### FULFILL
+
+- `fulfillment`: `0000000000000000000000000000000000000000000000000000000000000000`
+- `data`: _Response message from recipient settlement engine_
+
+##### REJECT
+
+- `code`: _Determined by connector from HTTP status of forwarded request_
+- `triggeredBy`: `peer.settle`
+- `message`: _Determined by connector_
+- `data`: _Response message from recipient settlement engine, or empty if antecedent failure_
+
+### Motivation
+
+Settlement engines supercede the [Ledger Plugin Interface (LPIv2)](../deprecated/0024-ledger-plugin-interface-2/0024-ledger-plugin-interface-2.md), an earlier abstraction for settlement integrations with Interledger. This new model addresses these issues:
+
+1. Multi-account plugins required logic for handling ILP packets, increasing implementation complexity.
+2. Plugins bundled settlement and bilateral communication functionality together, limiting composability.
+3. JavaScript plugins limited interoperability with non-JavaScript connector implementations.
+4. Plugins operated in the same process as the connector, which limited scaling the two services independently.
+
+## Specification
+
+### Accounts and identifiers
+
+Each account MUST be identified by a unique, [URL-safe](https://tools.ietf.org/html/rfc3986#section-2.3) string.
+
+The settlement engine MUST be responsible for correlating an account identifier to the peer's identity on the shared ledger or settlement system, if required. For separation of concerns between clearing and settlement, the accounting system is NOT RECOMMENDED to have knowledge of the peer's identity on the shared settlement system.
+
+### Units and quantities
+
+An amount of a given asset or currency may be represented using different denominations or units of the asset. For example, one U.S. dollar may be represented as \$1 or 100 cents, and one Bitcoin may be represented as 1 BTC or 100000000 satoshis.
```

Comment:
```suggestion
Interledger measures accounts using an asset identifier that can represent any type of fungible asset. These types include currencies or equities like `USD`, `BTC`, `XRP`, or `AAPL`, but can also be contrived for a specific use-case, such as `WIDGETS`.

Additionally, each account includes a unit of account for the asset being tracked, called an `amount`.

Asset amounts may be represented using different denominations. For example, one U.S. dollar may be represented as \$1 or 100 cents, each of which is an equivalent unit of value. Likewise, one Bitcoin may be represented as 1 BTC or 100,000,000 satoshis.
```

Reply:
I updated it with the last line you suggested, much clearer 👍

---

On asset codes: I'd prefer to not reference them in the spec, including not defining assets in terms of them and not including them in requests/responses.

My view is that even though different settlement systems may utilize the same asset code, they may represent functionally distinct assets.

Suppose we have ETH paychan, ETH on-ledger, Raiden, and Plasma integrations. In each case, the asset code may be "ETH," but they have very different properties. Amounts held in one cannot be directly transferred to another, and some (e.g. Plasma ETH) may not be fungible or could even have a different exchange rate compared to normal, base layer ETH!

Could we define assets in a more general way that doesn't link them to their code?

Reply:
Yeah, actually, I like that instinct -- no sense putting something into the RFC if we don't need it.

Reply:
Added a more explicit explanation for why the two systems may need different scales, [how about this](https://github.com/interledger/rfcs/compare/579e27cf5209cf3cda4bd19f615f647e3fd48776...18f2eec9c89295778b158070d6ad3e4b79dafb3e#diff-990a915317f4e35578edfa70631bd07dR199)?

Reply:
> Added a more explicit explanation for why the two systems may need different scales, how about this?

Sounds fine, although I would just delete line [201](https://github.com/interledger/rfcs/compare/579e27cf5209cf3cda4bd19f615f647e3fd48776...18f2eec9c89295778b158070d6ad3e4b79dafb3e#diff-990a915317f4e35578edfa70631bd07dR201). Line 203 stands on its own just fine, IMHO.

Reply:
I'm going to resolve this and redo my suggestion on the current version.